pilot/pkg/model/push_context.go

```diff
@@ -1588,26 +1587,33 @@ func (ps *PushContext) initServiceAccounts(env *Environment, services []*Service
 			if port.Protocol == protocol.UDP {
 				continue
 			}
+
 			var accounts sets.String
 			func() {
 				// First get endpoint level service accounts
 				shard, f := env.EndpointIndex.ShardsForService(string(svc.Hostname), svc.Attributes.Namespace)
 				if f {
 					shard.RLock()
-					defer shard.RUnlock()
-					accounts = shard.ServiceAccounts
+					// copy here to reduce the lock time, the copy cost can be ignored
```

Comment:
Why can it be ignored?

This seems less efficient; we always copy and before sometimes we avoided a copy.  Benchmark shows this as well (its faster because of the port change)

Reply:
Because nornally there are very small number of sas, so i say that